### PR TITLE
Make templates fallback methods template-agnostic

### LINKED ISSUE #42531: Hide the `Products by Category, Tag, and Attribute` templates from the templates list
Hide those three templates from the list of templates by default. Instead, they should appear in the `Add New` button menu. However, if the theme or the user has customized one of those three templates, it will appear in the templates list. (The tl;dr is that they should behave as the Category and Tag templates from WP core.)

### PR DESCRIPTION
**Pull request by Aljullu**
This PR:

* Gets rid of the `template_is_eligible_for_product_archive_fallback()` method, as knowing if a template is elligible for fallback is as easy as checking if `$template->template_fallback` exists.
* Refactors all `templates_is_eligible_for_product_archive_fallback...` methods to be template-agnostic, that means removing `product_archive` from the method name and checking for the `template_fallback` property, instead of only checking for the Product Catalog slug.

Sidenote: in the future we should be able to simplify this logic once we fix https://github.com/woocommerce/woocommerce/issues/42531, but there were some upstream blockers that didn't allow us to fix it yet (see https://github.com/woocommerce/woocommerce/pull/44778).

### How to test the changes in this Pull Request:

This PR doesn't change the behavior of how WooCommerce templates currently work in the Site Editor. Hopefully any issues would be evident (ie: a PHP fatal) or would be caught by e2e tests. In any case, to do some smoke testing, related to the templates fallback logic:

Context: the Products by Category, Products by Tag and Products by Attribute templates fall back to the Product Catalog template. That means that if there are customizations in the Product Catalog template, they are also reflected in the other templates as long as they haven't been customized as well.

1. Go to Appearance > Editor and make some edits to the Product Catalog and Products by Category template.
2. Go to the Shop page in the frontend (`/shop`) and verify the edits you did in the Product Catalog template are visible.
3. Go to a category page in the frontend (ie: `/product-category/clothing`) and verify the edits you did in the Products by Category template are visible.
4. Go to a tag page in the frontend (ie: `/product-tag/recommended`, note: you might need to create the tag first from Products > Tags) and verify the edits you did in the Product Catalog template are visible.
5. Revert the editions.
6. Visit the frontend pages again and verify there are no edits.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Blocks/Utils/BlockTemplateUtils.php [hunk 1]
```diff
@@ -510,15 +510,15 @@ public static function supports_block_templates( $template_type = 'wp_template'
...
 	 *  generally use the `archive-product` as a fallback if there are no specific overrides.
 	 *
 	 * @param string $template_slug Slug to check for fallbacks.
 	 * @return boolean
 	 */
-	public static function template_is_eligible_for_product_archive_fallback( $template_slug ) {
+	public static function template_is_eligible_for_fallback( $template_slug ) {
 		$registered_template = self::get_template( $template_slug );
 		if ( $registered_template && isset( $registered_template->fallback_template ) ) {
 			return ProductCatalogTemplate::SLUG === $registered_template->fallback_template;
```
Comment (gigitux) [inline plugins/woocommerce/src/Blocks/Utils/BlockTemplateUtils.php]: It seems like we’re currently only checking the fallback template for archive templates. Should we consider updating this check to cover other template types as well?​⬤

```
return ProductCatalogTemplate::SLUG === $registered_template->fallback_template;
```

Comment (gigitux) [reply plugins/woocommerce/src/Blocks/Utils/BlockTemplateUtils.php]: It looks like that we are not using this function. Should we remove it?

**Review by gigitux — APPROVED** [review]
The PR looks great! I've pre-approved it, but I left a comment regarding the `template_is_eligible_for_fallback` function. I believe we can remove this function as it may no longer be necessary.

Additionally, I suggest clarifying that any changes made to the Product Catalog are reflected in the Product Category template.

**Comment by Aljullu** [comment]
Thanks for the review, @gigitux!

> Additionally, I suggest clarifying that any changes made to the Product Catalog are reflected in the Product Category template.

I added an extra comment in 5807c497c27b8eba2d448a13863843f4d0369abd. Is that what you had in mind or would you add it somewhere else?

File: plugins/woocommerce/src/Blocks/Utils/BlockTemplateUtils.php (hunk 1 above)
Comment (Aljullu) [reply plugins/woocommerce/src/Blocks/Utils/BlockTemplateUtils.php]: Good catch, removed in 56fc3e6769454accc23b813449af4e7f48136e56.

**Comment by gigitux** [comment]
> Thanks for the review, @gigitux!
> 
> > Additionally, I suggest clarifying that any changes made to the Product Catalog are reflected in the Product Category template.
> 
> I added an extra comment in [5807c49](https://github.com/woocommerce/woocommerce/commit/5807c497c27b8eba2d448a13863843f4d0369abd). Is that what you had in mind or would you add it somewhere else?

Sorry for not being clear! I was referring to the testing instructions, but it is also a good point to add some details to the codebase.

:shipit:

**Comment by Aljullu** [comment]
> Sorry for not being clear! I was referring to the testing instructions, but it is also a good point to add some details to the codebase.

Aaah, ok! I added it to the testing instructions as well. :slightly_smiling_face:

**Pull request merged**